content/2022-02-25-pattern-matching-with-named-fields.md

```diff
@@ -0,0 +1,299 @@
+
+### Desugaring
+
+> One important principle of Scala’s pattern matching design is that case classes should be abstractable. I.e. we want to be able to represent the abilities of a case class without exposing the case class itself. That also allows code to evolve from a case class to a regular class or a different case class while maintaining the interface of the old case class. [Martin Odersky](https://contributors.scala-lang.org/t/pattern-matching-with-named-fields/1829/52)
+
+* Allow the same feature of case classes:
+  * Provide arbitrary names
+  * Enable [`@deprecatedName`](https://www.scala-lang.org/api/current/scala/deprecatedName.html), which leads to multiple names for a single field. Using both should be an error.
+* Maybe provide additional named fields. For example for our `User`:
+
+  ```scala
+  case User(ageInDays = 3650) => ...
+  ```
+
+  But this could be an awful feature.
+* Maybe leave fields without names
+
+This SIP proposes basically to use [shapeless records](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0#extensible-records) for the specific purpose of pattern matching.
+For an incomplete list of alternatives, see [alternative desugaring](#alternative-desugaring) below.
+
+Pro:
+
+* with shapeless records we have a good understanding what we are doing, however we need to tweak the encoding a bit
+* with some helper definitions, the usage could be quite short
+* reuse possible
+
+Con:
+
+* opens a can of worms, as this encoding could be useful for other requested features like [named tuples](named-tuple), which would require more thought-out design
+* no support for `@deprecatedName`
+* identifiers are represented with string literals
+* no meaningful way to use names within variadic patterns 
+  (the encoding enforces to use the names in order. A name like `last` in `Seq` could be implemented with this encoding.)
+
+Example for user:
+
+```scala
+object User:
+  def unapply(user: User): Option[
+    (String, Int, String) & 
+    { type Names = ("name", "age", "city") }
+  ] = Some((user.name, user.age, user.city)).asInstanceOf[
+      Option[(String, Int, String) & { type Names = ("name", "age", "city") }]
+    ]
+```
+
+// TODO: Describe why Names can also be use toplevel.
+
+The reason to pull the names out, instead of keeping them near to their type, is to prevent extractors to accidentally leaking the name.
+
+## Implementation
+
+If a pattern with a name is encountered, the compiler looks up the list of provided names and places the trees accordingly.
+
+The list of names either provided by the return type of the unapply method or by the constructor list of the case class.
+
+Example:
+
+```scala
+// a match clause like
+case User(<tree1>, city = <tree2>) => ???
+
+// gets rewritten to:
+case User(
+  <tree1>, // because tree1 is positional pattern
+  _, // because name isn't mentioned
+  <tree2>, // because city is the third parameter of user
+)
```

Comment:
I am not a reviewer but I would like to add that this solution would make it impossible to manually implement named field patterns.

Currently, it is possible to implement an `unapply` method in the companion object of a (non-case) class to implement an extractor supporting a constructor pattern:

~~~ scala
class Rational(val num: Int, val denom: Int)

object Rational:
  def unapply(rational: Rational): Option[(Int, Int)] = ...
~~~

However, with your solution, there does not seem to be a simple way to manually implement an extractor that supports named field patterns. This could be useful, e.g. for data types that can’t be case classes because they need to support evolutions in a binary compatible way. Or, maybe for extracting things from a Map:

~~~ scala
httpRequest match
  case GET -> Root / "foo" :? QueryParameter(q = query) => println(query)
~~~

Reply:
Maybe I got you wrong, but it is possible. Your example would be:
```
object Rational:
  def unapply(rational: Rational)
      : Option[(Int, Int) & { type Names = ("num", "denom") }] = ...
```

Regarding the possibility of extracting things from a map: I tried to keep it out of this SIP, but keep a way open to add it later - maybe. A hidden design goal, as people got confused - I think.

As I think about this now, the now proposed encoding doesn't fit that well in this regards. I'll add it to the cons list below. Thanks for bringing it up.

Reply:
> Maybe I got you wrong, but it is possible. Your example would be:
> 
> ```scala
> object Rational:
>   def unapply(rational: Rational)
>       : Option[(Int, Int) & { type Names = ("num", "denom") }] = ...
> ```

I think you understood my question well. However, it is not clear to me how, from this `unapply` signature, the compiler would be able to perform the rewrite you showed, which translates the whole thing into positional patterns.